### PR TITLE
Signature verification happy path with test

### PR DESCRIPTION
**Pull request by mbaynton**
Adds barebones versions of a bunch of lower-level operations that you need to perform a tuf signature verification (key id computation, json normalization, key and role organization in some kind of database) and performs one of the signature verifications, that timestamp.json is properly signed by a threshold of keys allowed to do so.

### CODE REVIEW

**Comment by heddn** [comment]
Based on the failing tests, we need to add the libsodium polyfill as a dependency for older versions of php.

**Comment by mbaynton** [comment]
Just getting back to sprinting in earnest today, yeah I'll drop the polyfill in now.

**Comment by heddn** [comment]
Still failing on hex2bin php 5.3, but maybe we can resolve that with jumping to php 5.5+

File: tests/Helpers/WebserverRunner.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
+<?php
```
Comment (tedbow) [inline tests/Helpers/WebserverRunner.php:1]: Should we just remove this?

File: src/Client/Updater.php [hunk 2]
```diff
@@ -11,13 +17,149 @@ class Updater
...
+        $this->mirrors = $mirrors;
+    }
+
+
+  /**
+   * @todo Update from python comment https://github.com/theupdateframework/tuf/blob/1cf085a360aaad739e1cc62fa19a2ece270bb693/tuf/client/updater.py#L999
+   *
+   * @param bool $unsafely_update_root_if_necessary
+   */
+    public function refresh($unsafely_update_root_if_necessary = true)
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/Updater.php:50]: Up to you, I think you know more about the intent than I. This is from straight porting the python?

File: src/Client/Updater.php [hunk 3]
```diff
@@ -11,13 +17,149 @@ class Updater
...
+        list($roleDb, $keyDb) = RepositoryDBCollection::singleton()->getDatabasesForRepository();
+        $keyMeta = $keyDb->getKey($signatureMeta['keyid']);
+        $pubkey = $keyMeta['keyval']['public'];
+        $pubkeyBytes = hex2bin($pubkey);
+        $sigBytes = hex2bin($signatureMeta['sig']);
+        // @TODO check that the key type in $signatureMeta is ed25519; return false if not.
+        return sodium_crypto_sign_verify_detached($sigBytes, $bytes, $pubkeyBytes);
+    }
+
+    private function getRepoFile($string)
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/Updater.php:156]:
```suggestion
    // To be replaced by HTTP / HTTP abstraction layer to the remote repository
    private function getRepoFile($string)
```

File: src/RoleDB.php [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    public static function createRoleDBFromRootMetadata($rootMetadata)
+    {
+        $db = new self();
+
+        foreach ($rootMetadata['roles'] as $roleName => $roleInfo) {
+            if ($roleName == 'root') {
+                $roleInfo['version'] = $rootMetadata['version'];
+                $roleInfo['expires'] = $rootMetadata['expires'];
+                $roleInfo['previous_keyids'] = $roleInfo['keyids'];
+                $roleInfo['previous_threshold'] = $roleInfo['threshold'];
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/RoleDB.php:24]: Not currently. I assume they have a purpose that will become apparent as we dive in more.
```suggestion
```

File: src/RoleDB.php [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+          $roleInfo['partial_loaded'] = false;
+          */
+
+            if (strncmp($roleName, 'targets', strlen('targets') === 0)) {
+                $roleInfo['paths'] = array();
+                $roleInfo['delegations'] = array('keys' => array(), 'roles' => array());
+            }
+            $roleInfo['paths'] = array();
+
+            if ($roleInfo['threshold'] + 0 <= 0) {
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/RoleDB.php:41]: Indifferent on that, both work the same right?

File: src/Client/Updater.php [hunk 6]
```diff
@@ -11,13 +17,149 @@ class Updater
...
+        $haveVerified = 0;
+
+        $canonical_bytes = JsonNormalizer::asNormalizedJson($signed);
+        foreach ($signatures as $signature) {
+            if ($this->isKeyIdAcceptableForRole($signature['keyid'], $type)) {
+                $haveVerified += (int)$this->verifySingleSignature($canonical_bytes, $signature);
+            }
+            if ($haveVerified >= $needVerified) {
+                break;
+            }
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/Updater.php:132]: That's my understanding, that you can have a key that straight up fails signature verification and it doesn't matter as long as the prescribed number by root.json do not fail. @davidstrauss can you confirm? Halp?

File: tests/Helpers/WebserverRunner.php (hunk 1 above)
Comment (mbaynton) [reply tests/Helpers/WebserverRunner.php:1]: Yes, I was going to do more changes in another PR though.

Comment (mbaynton) [reply tests/Helpers/WebserverRunner.php:1]: Oh did this never make it to `main` -- yeah okay I'll remove it

File: src/Client/Updater.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (davidstrauss) [reply src/Client/Updater.php:132]: I suspect we should ignore bad signatures as long as we're hitting thresholds. I actually worry that, if we don't, we might open an attack vector through an attacker providing a bad signature somehow (though I haven't looked at how).

Maybe we should warn on bad signatures, at least if we encounter them (which we may not in our sort of short-circuit evaluation)?

File: src/Client/Updater.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (tedbow) [reply src/Client/Updater.php:50]: Yeah I think we should remove it. We are doing a different approach now

File: src/RoleDB.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (tedbow) [reply src/RoleDB.php:39]: Sure it's fine

File: src/Client/Updater.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (tedbow) [reply src/Client/Updater.php:132]: Ok. I think at least a todo for logging then. Just seems like if this happening you would want to surface it. It seems like our (future)tests should also fail in this situation unless they were expecting the warning

[the remark this answers is not in the document]
Comment (tedbow) [reply src/Client/Updater.php:135]: Added a comment

File: src/RoleDB.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (xjm) [reply src/RoleDB.php:39]: Yeah this is weird; if this were in a core patch I'd ask for it to be changed.

**Pull request merged**